zrml/neo-swaps/src/lib.rs

```diff
@@ -0,0 +1,887 @@
+// Copyright 2023 Forecasting Technologies LTD.
+//
+// This file is part of Zeitgeist.
+//
+// Zeitgeist is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the
+// Free Software Foundation, either version 3 of the License, or (at
+// your option) any later version.
+//
+// Zeitgeist is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Zeitgeist. If not, see <https://www.gnu.org/licenses/>.
+
+#![doc = include_str!("../README.md")]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+
+mod benchmarking;
+mod consts;
+mod math;
+mod mock;
+mod tests;
+pub mod traits;
+pub mod types;
+pub mod weights;
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+mod pallet {
+    use crate::{
+        consts::MAX_ASSETS,
+        math::{Math, MathOps},
+        traits::{pool_operations::PoolOperations, DistributeFees, LiquiditySharesManager},
+        types::{FeeDistribution, Pool, SoloLp},
+        weights::*,
+    };
+    use alloc::{collections::BTreeMap, vec, vec::Vec};
+    use core::marker::PhantomData;
+    use frame_support::{
+        dispatch::DispatchResultWithPostInfo,
+        ensure,
+        pallet_prelude::StorageMap,
+        require_transactional,
+        traits::{Get, IsType, StorageVersion},
+        transactional, PalletId, Twox64Concat,
+    };
+    use frame_system::{ensure_signed, pallet_prelude::OriginFor};
+    use orml_traits::MultiCurrency;
+    use sp_runtime::{
+        traits::{AccountIdConversion, CheckedSub, Saturating, Zero},
+        DispatchError, DispatchResult, SaturatedConversion,
+    };
+    use zeitgeist_primitives::{
+        constants::{BASE, CENT},
+        math::fixed::{bdiv, bmul},
+        traits::{CompleteSetOperationsApi, DeployPoolApi},
+        types::{Asset, MarketStatus, MarketType, ScalarPosition, ScoringRule},
+    };
+    use zrml_market_commons::MarketCommonsPalletApi;
+
+    // These should not be config parameters to avoid misconfigurations.
+    pub(crate) const MIN_SWAP_FEE: u128 = BASE / 1_000; // 0.1%.
+    pub(crate) const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+    pub(crate) const MAX_SPOT_PRICE: u128 = BASE - CENT / 2;
+    pub(crate) const MIN_SPOT_PRICE: u128 = CENT / 2;
+    pub(crate) const MIN_LIQUIDITY: u128 = BASE;
+
+    pub(crate) type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+    pub(crate) type AssetOf<T> = Asset<MarketIdOf<T>>;
+    pub(crate) type BalanceOf<T> = <<T as Config>::MultiCurrency as MultiCurrency<
+        <T as frame_system::Config>::AccountId,
+    >>::Balance;
+    pub(crate) type IndexType = u32;
```

Comment:
Counts are always indices, hence the name. But I see the point about this not being the only index type in the pallet, so I've renamed it. Real question is, why am I using `u32` instead of `u16`? 4e38a45372d64285e8529bb1073aa8cb264e803a